auth/auth.go

```diff
@@ -0,0 +1,115 @@
+package auth
+
+import (
+	"crypto/ed25519"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/brave/go-sync/datastore"
+	"github.com/brave/go-sync/utils"
+	"github.com/satori/go.uuid"
+)
+
+var (
+	timestampMaxDuration int64 = 120 * 1e3 // Milliseconds, modifiable in tests.
+)
+
+const (
+	tokenMaxDuration int64  = 86400 * 1e3 // Milliseconds
+	bearerPrefix     string = "Bearer "
+)
+
+// Request is a struct used for authenication requests.
+type Request struct {
+	PublicKey       string `json:"public_key"`
+	Timestamp       string `json:"timestamp"`
+	SignedTimestamp string `json:"signed_timestamp"`
+}
+
+// Response is a struct used for authenication responses.
+type Response struct {
+	AccessToken string `json:"access_token"`
+	ExpiresIn   int64  `json:"expires_in"`
+}
+
+// Authenticate validates client's auth requests and provides the reply.
+func Authenticate(r *http.Request, db datastore.Datastore) (string, []byte, error) {
+	var rsp []byte
+
+	err := r.ParseForm()
+	if err != nil {
+		return "", nil, err
+	}
+	req := &Request{
+		PublicKey:       r.PostFormValue("client_id"),
+		Timestamp:       r.PostFormValue("timestamp"),
+		SignedTimestamp: r.PostFormValue("client_secret"),
+	}
+
+	// Verify the signature.
+	publicKey, err := hex.DecodeString(req.PublicKey)
+	if err != nil {
+		return "", nil, err
+	}
+	timestampBytes, err := hex.DecodeString(req.Timestamp)
+	if err != nil {
+		return "", nil, err
+	}
+	signedTimestamp, err := hex.DecodeString(req.SignedTimestamp)
+	if err != nil {
+		return "", nil, err
+	}
+	if !ed25519.Verify(publicKey, timestampBytes, signedTimestamp) {
+		return "", nil, fmt.Errorf("signature verification failed")
+	}
+
+	var timestamp int64
+	timestamp, err = strconv.ParseInt(string(timestampBytes), 10, 64)
+	if err != nil {
+		return "", nil, fmt.Errorf("parse timestamp error")
+	}
+
+	// Verify the timestamp is not outdated
+	if utils.UnixMilli(time.Now())-timestamp > timestampMaxDuration {
+		return "", nil, fmt.Errorf("timestamp is outdated")
+	}
+
+	// Create a new token, save it in DB, and return it.
+	expireAt := utils.UnixMilli(time.Now().Add(time.Duration(tokenMaxDuration) * time.Millisecond))
+	token := uuid.NewV4().String()
```

Comment:
i'm concerned about the use of UUID as an authentication token for sync operations (if i understand this code correctly), because according to https://www.ietf.org/rfc/rfc4122.txt:

>    Do not assume that UUIDs are hard to guess; they should not be used as security capabilities (identifiers whose mere possession grant access), for example. 

obviously this could vary depending on the UUID implmentation, but in general i think it's much safer to generate 32-bytes of randomness from a cryptographically random source and base64 or hex encode it. https://golang.org/pkg/crypto/rand/ works.

Reply:
looking at the implementation in https://github.com/satori/go.uuid/blob/d91630c8510268e75203009fe7daf2b8e1d60c45/generator.go#L168, it does appear to use golang's crypto rand package, which is good, because that means this is at least 122 bits of randomness. 

however:
1. the code here should throw an error if an error is thrown in https://github.com/satori/go.uuid/blob/d91630c8510268e75203009fe7daf2b8e1d60c45/generator.go#L168 because that could indicate a problem with the system random number generator, meaning tokens aren't actually random
2. it would still be better to be 256-bits instead of 122

Reply:
@diracdeltas Thanks for pointing this out, I will prepare a commit to generate 32 bytes randomness instead and make sure the error is handled when there is one from `rand.Read`.

Reply:
Just a side note here, we are using v1.2.0 of go.uuid which does not return error but panic directly when `rand.Read` has error, https://github.com/satori/go.uuid/blob/v1.2.0/generator.go#L200-L202.

Reply:
Addressed in https://github.com/brave/go-sync/pull/1/commits/8e95722f5b31f5aa770d39d0d36e606cf055931f

Reply:
nice, lgtm